src/sync/sys/unix/net.rs

```diff
@@ -0,0 +1,333 @@
+/*
+	Copyright The containerd Authors.
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+use crate::error::Result;
+use nix::sys::socket::*;
+use std::io::{self};
+use std::os::unix::io::RawFd;
+use std::os::unix::prelude::AsRawFd;
+use nix::Error;
+
+use nix::unistd::*;
+use std::sync::{Arc};
+use std::sync::atomic::{AtomicBool, Ordering};
+use crate::common::{self, client_connect, SOCK_CLOEXEC};
+#[cfg(target_os = "macos")] 
+use crate::common::set_fd_close_exec;
+use nix::sys::socket::{self};
+
+pub struct PipeListener {
+    fd: RawFd,
+    monitor_fd: (RawFd, RawFd),
+}
+
+impl AsRawFd for PipeListener {
+    fn as_raw_fd(&self) -> RawFd {
+        self.fd
+    }
+}
+
+impl PipeListener {
+    pub(crate) fn new(sockaddr: &str) -> Result<PipeListener> {
+        let (fd, _) = common::do_bind(sockaddr)?;
+        common::do_listen(fd)?;
+
+        let fds = PipeListener::new_monitor_fd()?;
+
+        Ok(PipeListener {
+            fd,
+            monitor_fd: fds,
+        })
+    }
+
+    pub(crate) fn new_from_fd(fd: RawFd) -> Result<PipeListener> {
+        let fds = PipeListener::new_monitor_fd()?;
+
+        Ok(PipeListener {
+            fd,
+            monitor_fd: fds,
+        })
+    }
+
+    fn new_monitor_fd() ->  Result<(i32, i32)> {
+        #[cfg(target_os = "linux")]
+        let fds = pipe2(nix::fcntl::OFlag::O_CLOEXEC)?;
+ 
+        
+        #[cfg(target_os = "macos")] 
+        let fds = {
+            let (rfd, wfd) = pipe()?;
+            set_fd_close_exec(rfd)?;
+            set_fd_close_exec(wfd)?;
+            (rfd, wfd)
+        };
+
+
+        Ok(fds)
+    }
+
+    pub(crate) fn accept( &self, quit_flag: &Arc<AtomicBool>) ->  std::result::Result<Option<PipeConnection>, io::Error> {
+        if quit_flag.load(Ordering::SeqCst) {
+            info!("listener shutdown for quit flag");
+            return Err(io::Error::new(io::ErrorKind::Other, "listener shutdown for quit flag"));
```

Comment:
Line 83-84 is a little confused for log with info but returns with an error.

Reply:
updated